cilium-health/launch/endpoint.go

```diff
@@ -0,0 +1,211 @@
+	if err != nil {
+		log.WithError(err).Fatal("Error while allocating cilium-health IP")
+	}
+
+	// XXX: Make the above IPs known to pkg/node
+
+	// XXX: Release IPs later on?
+
+	// Prepare the endpoint change request
+	id := int64(addressing.CiliumIPv6(ip6).EndpointID())
+	info := &models.EndpointChangeRequest{
+		ID:            id,
+		ContainerID:   endpoint.NewCiliumID(id),
+		ContainerName: "cilium-health",
+		State:         models.EndpointStateWaitingForIdentity,
+		Addressing: &models.EndpointAddressing{
+			IPV6: ip6.String(),
+			IPV4: ip4.String(),
+		},
+	}
+
+	// Create the veth pair for the health endpoint
+	vethName := "cilium_health"
+	vethPeerName := "cilium"
+	if link, err := netlink.LinkByName(vethName); err == nil {
+		netlink.LinkDel(link)
+	} else {
+		log.Debug("Didn't find existing link")
+	}
+	veth, _, err := plugins.SetupVethWithNames(vethName, vethPeerName, 1450, info)
+	if err != nil {
+		log.WithError(err).Fatal("Error while creating cilium-health veth")
+	}
+	defer func() {
+		if err == nil {
+			return
+		}
+		if err = netlink.LinkDel(veth); err != nil {
+			log.WithError(err).WithField(logfields.Veth, veth.Name).Warn("failed to clean up veth")
+		}
+	}()
+
+	// Invoke spawn_netns.sh to spin up the health endpoint
+	pidfile := fmt.Sprintf("%s/health-endpoint.pid", owner.GetStateDir())
+	os.RemoveAll(pidfile)
+	healthArgs := fmt.Sprintf("-d --admin=unix --passive --pidfile %s", pidfile)
+	args := []string{info.ContainerName, info.InterfaceName, vethPeerName,
+		ip6.String(), ip4.String(), "cilium-health", healthArgs}
+	prog := filepath.Join(owner.GetBpfDir(), "spawn_netns.sh")
+
+	ctx, cancelHealth := context.WithCancel(context.Background())
+	cmd := exec.CommandContext(ctx, prog, args...)
+	if err = logFromCommand(cmd, info.ContainerName); err != nil {
+		log.WithError(err).Fatal("Error while opening pipes to health endpoint")
+	}
+	if err = cmd.Start(); err != nil {
+		target := fmt.Sprintf("%s %s", prog, strings.Join(args, " "))
+		log.WithField("cmd", target).WithError(err).Fatal("Error spawning cilium-health endpoint")
+	}
+
+	// Create the endpoint
+	lbl := labels.Labels{labels.IDNameHealth: labels.NewLabel(labels.IDNameHealth, "", labels.LabelSourceReserved)}
+	ep, err := endpoint.NewEndpointFromChangeModel(info, lbl)
+	if err != nil {
+		log.WithError(err).Fatal("Error while creating cilium-health endpoint")
+	}
+	ep.SetDefaultOpts(opts)
+
+	// Add the endpoint
+	if err := endpointmanager.AddEndpoint(owner, ep, "Create cilium-health endpoint"); err != nil {
+		log.WithError(err).Fatal("Error while adding cilium-health endpoint")
+	}
+
+	// Give the endpoint a security identity
+	_, errLabelsAdd := owner.UpdateSecLabels(info.ContainerID, lbl, labels.Labels{})
+	if errLabelsAdd != nil {
+		log.WithError(errLabelsAdd).Fatal("Failed to set labels on cilium-health endpoint")
+	}
+
+	// Wait until the cilium-health endpoint is running before setting up routes
+	deadline := time.Now().Add(5 * time.Second)
```

Comment:
Are 5 seconds enough?

Reply:
It's pretty arbitrary, do you have another number in mind?

By the way, if this fails then it will only stall cilium daemon startup by max 5 seconds so in some ways it's nice -- we should never expect to hit this but if we do, then at least it only affects the health endpoint.